numpy/core/src/multiarray/datetime_strings.c

```diff
@@ -1592,3 +1593,316 @@ array_datetime_as_string(PyObject *NPY_UNUSED(self), PyObject *args,
+        if (PyBytes_Check(timezone_obj)) {
+            char *str;
+            Py_ssize_t len;
+
+            if (PyBytes_AsStringAndSize(timezone_obj, &str, &len) < 0) {
+                goto fail;
+            }
+            else {
+                PyErr_Format(PyExc_ValueError, "Unsupported timezone "
+                            "input string \"%s\"", str);
+                goto fail;
+            }
+        }
+    }
+
+    /* Get a string size long enough for any datetimes we're given
+    note we can't really know this ahead of time unless we format
+    all the strings because the year field can be arbitrarily
+    long */
+    strsize = strlen(fmt) + 20;
+#if defined(NPY_PY3K)
+    /*
+     * For Python3, allocate the output array as a UNICODE array, so
+     * that it will behave as strings properly
+     */
+    op_dtypes[1] = PyArray_DescrNewFromType(NPY_UNICODE);
+    if (op_dtypes[1] == NULL) {
+        goto fail;
+    }
+    op_dtypes[1]->elsize = strsize * 4;
+    /* This steals the UNICODE dtype reference in op_dtypes[1] */
+    op[1] = (PyArrayObject *)PyArray_NewLikeArray(op[0],
+                                        NPY_KEEPORDER, op_dtypes[1], 1);
+    if (op[1] == NULL) {
+        op_dtypes[1] = NULL;
+        goto fail;
+    }
+#endif
+    /* Create the iteration string data type (always ASCII string) */
+    op_dtypes[1] = PyArray_DescrNewFromType(NPY_STRING);
+    if (op_dtypes[1] == NULL) {
+        goto fail;
+    }
+    op_dtypes[1]->elsize = strsize;
+
+    flags = NPY_ITER_ZEROSIZE_OK|
+            NPY_ITER_BUFFERED;
+    op_flags[0] = NPY_ITER_READONLY|
+                  NPY_ITER_ALIGNED;
+    op_flags[1] = NPY_ITER_WRITEONLY|
+                  NPY_ITER_ALLOCATE;
+
+    iter = NpyIter_MultiNew(2, op, flags, NPY_KEEPORDER, NPY_UNSAFE_CASTING,
+                            op_flags, op_dtypes);
+    if (iter == NULL) {
+        goto fail;
+    }
+
+    if (NpyIter_GetIterSize(iter) != 0) {
+        NpyIter_IterNextFunc *iternext;
+        char **dataptr;
+        npy_datetime dt;
+        npy_datetimestruct dts;
+
+        iternext = NpyIter_GetIterNext(iter, NULL);
+        if (iternext == NULL) {
+            goto fail;
+        }
+        dataptr = NpyIter_GetDataPtrArray(iter);
+
+        do {
+            int tzoffset = -1;
+
+            /* Get the datetime */
+            dt = *(npy_datetime *)dataptr[0];
+
+            /* Convert it to a struct */
+            if (convert_datetime_to_datetimestruct(meta, dt, &dts) < 0) {
+                goto fail;
+            }
+
+            /* Get the tzoffset from the timezone if provided */
+            tzoffset = -1;
+            if (timezone_obj != NULL) {
+                tzoffset = get_tzoffset_from_pytzinfo(timezone_obj, &dts);
+            }
+
+            if (tzoffset != -1){
+              add_minutes_to_datetimestruct(&dts, tzoffset);
+            }
+
+            /* Zero the destination string completely */
```

Comment:
Can you extract everything from this point forward into a helper function? The indentation is getting very deep here, although indenting with two instead of four spaces was a good attempt at making it less noticeable.

Reply:
Done